testnet/stacks-node/src/burnchains/db_indexer.rs

```
@@ -0,0 +1,589 @@
+            }
+        };
+
+        up_cursor = cursor_header.parent_header_hash;
+    };
+
+    // Step 2: Set `is_canonical` to false from the old tip (inclusive) to the greatest
+    // common ancestor (exclusive).
+    let mut down_cursor = BurnchainHeaderHash(old_tip.header_hash());
+    loop {
+        let cursor_header = match query_row::<BurnHeaderDBRow, _>(
+            connection,
+            "SELECT * FROM headers WHERE header_hash = ?1",
+            &[&down_cursor],
+        )? {
+            Some(header) => header,
+            None => {
+                // TODO: Should this be an error?
+                panic!("Do we hit here?");
+            }
+        };
+
+        if cursor_header.header_hash == greatest_common_ancestor.header_hash {
+            break;
+        }
+
+        match connection.execute(
+            "UPDATE headers SET is_canonical = 0 WHERE header_hash = ?1",
+            &[&down_cursor],
+        ) {
+            Ok(_) => {}
+            Err(e) => {
+                return Err(BurnchainError::DBError(db_error::SqliteError(e)));
+            }
+        };
+
+        down_cursor = cursor_header.parent_header_hash;
+    }
+
+    Ok(greatest_common_ancestor.height)
+}
+
+/// Returns the first ancestor of `last_canonical_tip` that is marked canonical. After a re-org, this
+/// can be used to find the greatest common ancestor between the new and old chain tips.
+fn find_first_canonical_ancestor(
+    connection: &DBConn,
+    last_canonical_tip: BurnchainHeaderHash,
+) -> Result<u64, BurnchainError> {
+    let mut cursor = last_canonical_tip;
+    loop {
+        let cursor_header = match query_row::<BurnHeaderDBRow, _>(
+            connection,
+            "SELECT * FROM headers WHERE header_hash = ?1",
+            &[&cursor],
+        )? {
+            Some(header) => header,
+            None => {
+                // TODO: Should this be an error?
+                panic!("Do we hit here?");
```

Yes, this should throw an error -- this shouldn't happen (no parent for a block should be absent), but this method returns a `Result` anyways, so we should `error!` log and then return an error variant here.

---

But, this would be an internal error, I think, and don't we usually panic on those?

I mean, it's an "internal error" because we should only be processing a reorg if 1) the parent of `new_tip` exists, 2) `old_tip` exists.. and also, all the ancestors of these should exist, and the root of the tree should be `is_canonical`.

If so, I can update the contract to explain that these are expected to exist.

---

It's only kind of internal -- it may indicate an error in the Stacks node, or an error in the deployment because an event was missed. It isn't recoverable yet, so panicking is probably what the caller of this method will do, but there's no real reason not to let the caller make that decision. When we introduce support for querying older block commits (rather than simply digesting events), this _would_ be recoverable, so it makes sense to at least produce the error variant here, and let a higher level function `.expect()` the result.

In any event, this branch should _definitely_ produce an `error!` log with information about what the error is.